File: python/detectionformats/source.py
#!/usr/bin/env python

#stdlib imports
import json


class Source:
    """ Source - a conversion class used to create, parse, and validate source 
        data as part of detection data.
    """
    # json keys
    AGENCYID_KEY = "AgencyID"
    AUTHOR_KEY = "Author"

    def __init__(self, newAgencyID=None, newAuthor=None):
        """Initialize the source object. Constructs an empty object
           if all arguments are None

        Args:
            newAgencyID: a required String containing the agency identifier
            newAuthor: a required String containing the author name
        Returns:
            Nothing
        Raises:
            Nothing
        """
        if newAgencyID is not None:
            self.agencyID = newAgencyID
        if newAuthor is not None:
            self.author = newAuthor

    def fromJSONString(self, jsonString):
        """Populates the object from a json formatted string

        Args:
            jsonString: a required String containing the json formatted text
        Returns:
            Nothing
        Raises:
            Nothing
        """
        jsonObject = json.loads(jsonString)
        self.fromDict(jsonObject)

    def fromDict(self, aDict):
        """Populates the object from a dictionary

        Args:
            aDict: a required dictionary
        Returns:
            Nothing
        Raises:
            Nothing
        """
        try:
            self.agencyID = aDict[self.AGENCYID_KEY]
            self.author = aDict[self.AUTHOR_KEY]

        except (ValueError, KeyError, TypeError):
            print ("Dict format error")

    def toJSONString(self):
        """Converts the object to a json formatted string

        Args:
            None
        Returns:
            The JSON formatted message as a String
        Raises:
            Nothing
        """
        jsonObject = self.toDict()

        return json.dumps(jsonObject, ensure_ascii=False)

    def toDict(self):
        """Converts the object to a dictionary

        Args:
            None
        Returns:
            The dictionary
        Raises:
            Nothing
        """
        aDict = {}
        try:
            aDict[self.AGENCYID_KEY] = self.agencyID
            aDict[self.AUTHOR_KEY] = self.author
        except NameError:
            print ("Missing data error")

        return aDict

    def isValid(self):
        """Checks to see if the object is valid

        Args:
            None
        Returns:
            True if the object is valid, False otherwise
        Raises:
            Nothing
        """
        errorList = self.getErrors()

        if len(errorList) == 0:
            return True
        else:
            return False

    def getErrors(self):
        """Gets a list of object validation errors

        Args:
            None
        Returns:
            A List of Strings containing the validation error messages
        Raises:
            Nothing
        """
        errorList = []

        try:
            if self.agencyID == '':
                errorList.append('Empty AgencyID in Source Class.')
        except (NameError, AttributeError):
            errorList.append('No AgencyID in Source Class.')

        try:
            if self.author == '':
                errorList.append('Empty Author in Source Class.')
        except (NameError, AttributeError):
            errorList.append('No Author in Source Class.')

        return errorList

    def isEmpty(self):
        """Checks to see if the object is empty

        Args:
            None
        Returns:
            True if the object is empty, False otherwise
        Raises:
            Nothing
        """
        if hasattr(self, 'agencyID'):
            return False

        if hasattr(self, 'author'):
            return False

        return True

File: java/src/main/java/gov/usgs/detectionformats/Source.java
package gov.usgs.detectionformats;

import java.util.ArrayList;
import org.json.simple.JSONObject;

/**
 * a conversion class used to create, parse, and validate source data as part of detection data.
 *
 * @author U.S. Geological Survey &lt;jpatton at usgs.gov&gt;
 */
public class Source implements DetectionInt {

  /** JSON Keys */
  public static final String AGENCYID_KEY = "AgencyID";

  public static final String AUTHOR_KEY = "Author";

  /** Required Agency identifier */
  private final String agencyID;

  /** Required Author identifier */
  private final String author;

  /** The constructor for the Source class. Initializes members to null values. */
  public Source() {
    agencyID = null;
    author = null;
  }

  /**
   * The advanced constructor for the Source class. Initializes members to provided values.
   *
   * @param newAgencyID - A String containing the agencyID to use
   * @param newAuthor - A String containing the author to use
   */
  public Source(String newAgencyID, String newAuthor) {
    agencyID = newAgencyID;
    author = newAuthor;
  }

  /**
   * Constructs the class from a JSONObject, populating members
   *
   * @param newJSONObject - A JSONObject.
   */
  public Source(JSONObject newJSONObject) {
    // required values
    // agencyID
    if (newJSONObject.containsKey(AGENCYID_KEY)) {
      agencyID = newJSONObject.get(AGENCYID_KEY).toString();
    } else {
      agencyID = null;
    }

    // author
    if (newJSONObject.containsKey(AUTHOR_KEY)) {
      author = newJSONObject.get(AUTHOR_KEY).toString();
    } else {
      author = null;
    }
  }

  /**
   * Converts the contents of the class to a JSONObject
   *
   * @return Returns a JSONObject containing the class contents
   */
  @SuppressWarnings("unchecked")
  public JSONObject toJSON() {
    JSONObject newJSONObject = new JSONObject();

    String jsonAgencyID = getAgencyID();
    String jsonAuthor = getAuthor();

    // required values
    // agencyID
    if ((jsonAgencyID != null) && (!jsonAgencyID.isEmpty())) {
      newJSONObject.put(AGENCYID_KEY, jsonAgencyID);
    }

    // author
    if ((jsonAuthor != null) && (!jsonAuthor.isEmpty())) {
      newJSONObject.put(AUTHOR_KEY, jsonAuthor);
    }

    return (newJSONObject);
  }

  /**
   * Validates the class.
   *
   * @return Returns true if successful
   */
  public boolean isValid() {
    if (getErrors().size() == 0) {
      return (true);
    } else {
      return (false);
    }
  }

  /**
   * Gets any validation errors in the class.
   *
   * @return Returns a List&lt;String&gt; of any errors found
   */
  public ArrayList<String> getErrors() {
    String jsonAgencyID = getAgencyID();
    String jsonAuthor = getAuthor();

    ArrayList<String> errorList = new ArrayList<String>();

    // check for required keys
    // agencyID
    if (jsonAgencyID == null) {
      // agencyID not found
      errorList.add("No AgencyID in Source Class.");
    } else if (jsonAgencyID.isEmpty()) {
      // agencyID empty
      errorList.add("Empty AgencyID in Source Class.");
    }

    // author
    if (jsonAuthor == null) {
      // author not found
      errorList.add("No Author in Source Class.");
    } else if (jsonAuthor.isEmpty()) {
      // author empty
      errorList.add("Empty Author in Source Class.");
    }

    // success
    return (errorList);
  }

  /**
   * Checks to see if this object is empty
   *
   * @return Returns true if empty, false otherwise.
   */
  public boolean isEmpty() {
    if (getAgencyID() != null) {
      return (false);
    }
    if (getAuthor() != null) {
      return (false);
    }
    return (true);
  }

  /** @return the agencyID */
  public String getAgencyID() {
    return agencyID;
  }

  /** @return the author */
  public String getAuthor() {
    return author;
  }
}
